### ISSUE TITLE
GPU/CPU f32 APR forward diverges at layer-0 QKV (std 4.44 vs 0.30); gpu_cpu_trace_compare fails 91.9% L2 once it compiles

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3975]
Found while fixing #3956. Once `crates/aprender-serve/tests/gpu_cpu_trace_compare.rs` compiles again under `--features cuda`, it runs and FAILS:

```
PMAT-216 FAIL: GPU L2 diverged 91.92% from CPU (max 1%)
```

Measured on lambda (RTX 4090) under gpu-q, base `release/0.69.1-batch-2@fc942f6be` plus the one-line `lm_head_tied: false` compile fix.

**Where it splits:** at layer 0 QKV, not the lm_head. Embedding stats match (0.00%). I added temporary per-stage std prints (not committed):

| stage | CPU std | GPU std |
|---|---|---|
| L0 qkv | 4.43868 | 0.29929 |
| L0 attn_out | 1.45359 | 0.08645 |
| L0 ffn_out | 0.09265 | 0.00037 |
| L0 out | 1.47782 | 0.08677 |
| final_norm | 0.86840 | 0.09944 |
| logits | 23.3227 | 0.5123 |

The test's own layer table compares MEANS only and printed `qkv diff: 3.56% | out diff: 0.52%`, which hid a 15x std gap.

**Hypothesis (not proven):** the fixture's weights are `sin(i*0.01)*0.1`, which is smooth along a row. Read as `[out,in]`, each output is a coherent sum (large std); read transposed, it is incoherent (small std). So the CPU `forward_traced` (`helpers::f32_matmul`) and the GPU path (`AprF32ToGpuAdapter::convert_layer` transposes qkv to `[hidden, qkv_out]`, then `GpuModel` `do_matmul`) appear to disagree on QKV layout. Before blaming either side, this needs:
1. an element-wise check of the L0 qkv output against a hand-computed reference, not a std;
2. a second fixture (e.g. random weights) to rule out a fixture artifact (Rule 6);
3. a check of whether production `.apr` f32 → GPU serving reaches this adapter.

This test only builds with `--features cuda`, and cuda does not run in CI, so the divergence has been dark since at least the `lm_head_tied` field landed.

**Comment by noahgift** [comment]
**Root cause: the m=1 GEMV fast path in `CudaExecutor::gemm` reads B as `[n,k]`, while `gemm` documents B as `[k,n]`.** My layout hypothesis in the issue body (a transpose in the adapter) is refuted.

Evidence, on `release/0.69.1-batch-2@204abee3c` with the layer-0 QKV instrumented (scratch only, reverted), run under `scripts/lib/gpu_exclusive_run.sh`: rc 101, and the only process on the card was the test binary.

| | first QKV elements |
|---|---|
| GPU `do_matmul` → `CudaScheduler::matmul` → `executor.gemm(m=1)` | `0.2932977, -0.30121523, 0.29412347, -0.3016835` |
| CPU, same normed input × the same uploaded `[k,n]` buffer | `1.829577, 5.044075, 6.262087, 5.001509` |
| numpy, reading that uploaded buffer as `[n,k]` | `0.29329768, -0.30121523, 0.2941234, -0.3016835` |

- The CPU AprTransformer is correct: its trace qkv std is 4.439, and a numpy oracle on the fixture formulas in the contract layout `W[out,in]` gives 4.421.
- `AprF32ToGpuAdapter` is correct: it transposes to `[k,n]` as `gemm` documents.
- The fault is in `gemm_tiled.rs`: when `m == 1` it swaps in `KernelType::Gemv`, which reads `[n,k]`. `GemmTiled`, used for `m > 1`, reads `[k,n]`.

**The existing guard is vacuous.** `gemm_tests.rs::test_gemm_m1_uses_gemv` asserts only `c[0] > 0.0`, and its comment's expected value (3.25) is the misread result. The `[k,n]` answer is 9.5.

**Callers disagree about the layout:**
- `GpuModel` via `CudaScheduler` passes `[k,n]`: wrong at m=1 (this bug).
- `gguf/inference/matmul_fused.rs` (the dequant + `gemm` fallback) and `gguf/inference/cached/thread-safe.rs::batch_matmul_gpu_prefer_cuda` pass a dequantized `[out,in]` = `[n,k]` weight: right at m=1 only because of this quirk, wrong at m>1.

**Reachability:** `AprF32ToGpuAdapter` has no production caller, and I found no non-test `enable_cuda()` call. Whether the cached batch path serves a released cell has not been checked yet.

The fix scope is waiting on a decision from the cop.

**Comment by noahgift** [comment]
**Fix-design quorum: do-not-implement-as-written 2/2** on "force [k,n] everywhere" (gemini-3.1-pro-high, gemini-3.8-flash-high). **Adopted: Alternative 2, an explicit layout at the API** (`Layout::KN|NK`, or gemm + gemm_bt, matching `GpuExecutorTrait::matmul_transpose_b`). NK at m=1 keeps the coalesced Gemv, so decode speed is unaffected. Callers: (a) CudaScheduler = KN. (c) `batch_matmul_gpu_prefer_cuda` = NK; it is REACHABLE via `apr serve --gpu --batch` with an MHA model and ≥32 prompts, and today it computes garbage at m>1. (b) `fused_matmul_cuda` is dead code and gets deleted. Acceptance: aprender-19 [337148]'s 5 must-REDs (m=1 AND m=2 for every caller) at fix/3975-qkv-divergence@9ce6080a6, which **must execute on a GPU lane**; a skip on a GPU-less runner is not a gate.

**Comment by noahgift** [comment]
Fixed on `fix/3975-qkv-divergence@d1d229c93`, ready for the cop to fold.

- **`gemm` fix (quorum: explicit layout at the API):** `gemm` reads B as `[k,n]` at every m. The new `gemm_bt` reads `[n,k]`: m=1 uses the existing Gemv; m>1 uses a tiled kernel emitted by trueno's existing `GemmBackwardAKernel`. Every `[out,in]` caller is rerouted, including the HybridScheduler fallbacks, which were wrong at every m. The dead `fused_matmul_cuda` is deleted.
- **Second defect, needed for the test to pass:** `AprF32ToGpuAdapter` wrapped the `LmHeadWeight` / `LmHeadWeightTransposed` newtypes the wrong way round, and `batch.rs` consumed that swapped form. The old m=1 quirk cancelled the swap out. Both now follow the contract in `scheduler/types.rs`.
- **Evidence, lambda RTX 4090, exclusive card:**
  - `gpu_cpu_trace_compare`: GREEN.
  - Must-RED tests (`gguf::cuda::gemm_layout_tests_3975`): 6/6, in both debug and release.
  - Planted mutant (n/k arguments swapped in the `gemm_bt` launch): RED 2/6.
  - The tests live under `gguf::cuda::`, which CI's `cuda-unit` lane selects.
- **Follow-up:** `driver_cuda_gguf.rs` does not compile under cuda. Filed as #3989.

**Comment by noahgift** [comment]
milestone → 0.71.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: named must-carry in epic #3994
